### PR TITLE
octave: migrate from homebrew/science

### PR DESCRIPTION
**Pull request by fxcoudert**
See https://github.com/Homebrew/homebrew-core/issues/15619

### CODE REVIEW

File: Formula/octave.rb [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  depends_on "qrupdate"
+  depends_on "readline"
+  depends_on "suite-sparse"
+  depends_on "transfig"
+  depends_on "veclibfort"
+
+  # If GraphicsMagick was built from source, it is possible that it was
+  # done to change quantum depth. If so, our Octave bottles are no good.
+  # https://github.com/Homebrew/homebrew-science/issues/2737
+  if build.with? "graphicsmagick"
```
Comment (ilovezfs) [inline Formula/octave.rb]: this will alway be false since graphicsmagick is non-optional

File: Formula/octave.rb [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      # Remove for > 4.2.1
+      # Remove inline keyword on file_stat destructor which breaks macOS
+      # compilation (bug #50234).
+      # Upstream commit from 24 Feb 2017 http://hg.savannah.gnu.org/hgweb/octave/rev/a6e4157694ef
+      inreplace "liboctave/system/file-stat.cc",
+        "inline file_stat::~file_stat () { }", "file_stat::~file_stat () { }"
+    end
+
+    # basic arguments
+    args = %W[--prefix=#{prefix}
```
Comment (ilovezfs) [inline Formula/octave.rb]: can inline all of these since nothing is optionally appended to the array

File: Formula/octave.rb [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    # be inserted into every oct/mex build. This is actually unnecessary and
+    # can cause linking problems.
+    inreplace "src/mkoctfile.in.cc", /%OCTAVE_CONF_OCT(AVE)?_LINK_(DEPS|OPTS)%/, '""'
+
+    system "./configure", *args
+    system "make", "all"
+    system "make", "install"
+  end
+
+  def caveats; <<-EOS.undent
```
Comment (ilovezfs) [inline Formula/octave.rb]: I think we can just remove these caveats

File: Formula/octave.rb [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  end
+
+  depends_on "bison" => :build
+  depends_on "gawk" => :build
+  depends_on "gnu-sed" => :build # http://lists.gnu.org/archive/html/octave-maintainers/2016-09/msg00193.html
+  depends_on "pkg-config" => :build
+  depends_on "texinfo" => :build
+
+  depends_on :fortran
+  depends_on :x11
```
Comment (ilovezfs) [inline Formula/octave.rb]: can this be removed or at least made optional?

File: Formula/octave.rb [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    depends_on "icoutils" => :build
+    depends_on "librsvg" => :build
+  end
+
+  depends_on "bison" => :build
+  depends_on "gawk" => :build
+  depends_on "gnu-sed" => :build # http://lists.gnu.org/archive/html/octave-maintainers/2016-09/msg00193.html
+  depends_on "pkg-config" => :build
+  depends_on "texinfo" => :build
+
```
Comment (ilovezfs) [inline Formula/octave.rb]: shouldn't be spaces between dependency types

File: Formula/octave.rb [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  desc "High-level interpreted language for numerical computing"
+  homepage "https://www.gnu.org/software/octave/index.html"
+
+  stable do
+    url "https://ftp.gnu.org/gnu/octave/octave-4.2.1.tar.gz"
+    mirror "https://ftpmirror.gnu.org/octave/octave-4.2.1.tar.gz"
+    sha256 "80c28f6398576b50faca0e602defb9598d6f7308b0903724442c2a35a605333b"
+  end
+
+  head do
```
Comment (ilovezfs) [inline Formula/octave.rb]: let's just remove head

File: Formula/octave.rb [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    def pour_bottle?
+      Tab.for_name("graphicsmagick").bottle?
+    end
+  end
+
+  def install
+    ENV.m64 if MacOS.prefer_64_bit?
+    ENV.append_to_cflags "-D_REENTRANT"
+    ENV.append "LDFLAGS", "-L#{Formula["readline"].opt_lib} -lreadline"
+    ENV.prepend_path "PATH", Formula["texinfo"].bin
```
Comment (ilovezfs) [inline Formula/octave.rb]: superenv should already have done this for us

File: Formula/octave.rb [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+  if build.with? "graphicsmagick"
+    def pour_bottle?
+      Tab.for_name("graphicsmagick").bottle?
+    end
+  end
+
+  def install
+    ENV.m64 if MacOS.prefer_64_bit?
+    ENV.append_to_cflags "-D_REENTRANT"
+    ENV.append "LDFLAGS", "-L#{Formula["readline"].opt_lib} -lreadline"
```
Comment (ilovezfs) [inline Formula/octave.rb]: is it not able to find this on its own. superenv should take care of it

File: Formula/octave.rb [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+  # https://github.com/Homebrew/homebrew-science/issues/2737
+  if build.with? "graphicsmagick"
+    def pour_bottle?
+      Tab.for_name("graphicsmagick").bottle?
+    end
+  end
+
+  def install
+    ENV.m64 if MacOS.prefer_64_bit?
+    ENV.append_to_cflags "-D_REENTRANT"
```
Comment (ilovezfs) [inline Formula/octave.rb]: Is this a bug? Hopefully we can just remove it.

File: Formula/octave.rb [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  # done to change quantum depth. If so, our Octave bottles are no good.
+  # https://github.com/Homebrew/homebrew-science/issues/2737
+  if build.with? "graphicsmagick"
+    def pour_bottle?
+      Tab.for_name("graphicsmagick").bottle?
+    end
+  end
+
+  def install
+    ENV.m64 if MacOS.prefer_64_bit?
```
Comment (ilovezfs) [inline Formula/octave.rb]: don't think we need to worry about 32-bit

File: Formula/octave.rb [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    end
+
+    # basic arguments
+    args = %W[--prefix=#{prefix}
+              --enable-dependency-tracking
+              --enable-link-all-dependencies
+              --enable-shared
+              --disable-static
+              --disable-docs
+              --with-x=no
```
Comment (ilovezfs) [inline Formula/octave.rb]: with-x=no so that x11 dep is highly suspicious

File: Formula/octave.rb [hunk 12]
```diff
@@ -0,0 +1,136 @@
+class Octave < Formula
+  desc "High-level interpreted language for numerical computing"
+  homepage "https://www.gnu.org/software/octave/index.html"
+
+  stable do
```
Comment (ilovezfs) [inline Formula/octave.rb]: no need for a stable block

File: Formula/octave.rb [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+  depends_on "freetype"
+  depends_on "ghostscript"
+  depends_on "gl2ps"
+  depends_on "glpk"
+  depends_on "gnuplot"
+  depends_on "graphicsmagick"
+  depends_on "hdf5"
+  depends_on "libsndfile"
+  depends_on "metis"
+  depends_on "pcre"
```
Comment (ilovezfs) [inline Formula/octave.rb]: is macOS's pcre no good?

File: Formula/octave.rb [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+  depends_on "gawk" => :build
+  depends_on "gnu-sed" => :build # http://lists.gnu.org/archive/html/octave-maintainers/2016-09/msg00193.html
+  depends_on "pkg-config" => :build
+  depends_on "texinfo" => :build
+
+  depends_on :fortran
+  depends_on :x11
+
+  depends_on "arpack"
+  depends_on "curl"
```
Comment (ilovezfs) [inline Formula/octave.rb]: curl dep shouldn't be needed

File: Formula/octave.rb [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+              --without-qt
+              --with-blas=-L#{Formula["veclibfort"].opt_lib}\ -lvecLibFort
+              --with-fltk-prefix=#{Formula["fltk"].opt_prefix}
+              --with-hdf5-includedir=#{Formula["hdf5"].opt_include}
+              --with-hdf5-libdir=#{Formula["hdf5"].opt_lib}
+              --with-portaudio
+              --with-sndfile
+	     ]
+
+    ENV.append "LDFLAGS", "-L#{Formula["suite-sparse"].opt_lib} -lsuitesparseconfig"
```
Comment (ilovezfs) [inline Formula/octave.rb]: does superenv not do this for us?

File: Formula/octave.rb [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+    ENV.append "LDFLAGS", "-L#{Formula["metis"].opt_lib} -lmetis"
```
Comment (ilovezfs) [inline Formula/octave.rb]: does superenv not do this for us?

File: Formula/octave.rb [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+    depends_on "automake" => :build
+    depends_on "icoutils" => :build
+    depends_on "librsvg" => :build
+  end
+
+  depends_on "bison" => :build
+  depends_on "gawk" => :build
+  depends_on "gnu-sed" => :build # http://lists.gnu.org/archive/html/octave-maintainers/2016-09/msg00193.html
+  depends_on "pkg-config" => :build
+  depends_on "texinfo" => :build
```
Comment (ilovezfs) [inline Formula/octave.rb]: macOS's isn't sufficient?

File: Formula/octave.rb [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+  head do
+    url "http://www.octave.org/hg/octave", :branch => "default", :using => :hg
+    depends_on :hg => :build
+    depends_on "autoconf" => :build
+    depends_on "automake" => :build
+    depends_on "icoutils" => :build
+    depends_on "librsvg" => :build
+  end
+
+  depends_on "bison" => :build
```
Comment (ilovezfs) [inline Formula/octave.rb]: macOS's isn't sufficient?

File: Formula/octave.rb [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+  depends_on "gawk" => :build
```
Comment (ilovezfs) [inline Formula/octave.rb]: macOS's awk isn't sufficient?

File: Formula/octave.rb [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+              --enable-dependency-tracking
+              --enable-link-all-dependencies
+              --enable-shared
+              --disable-static
+              --disable-docs
+              --with-x=no
+              --without-OSMesa
+              --without-qt
+              --with-blas=-L#{Formula["veclibfort"].opt_lib}\ -lvecLibFort
+              --with-fltk-prefix=#{Formula["fltk"].opt_prefix}
```
Comment (ilovezfs) [inline Formula/octave.rb]: doesn't find this on its own?

File: Formula/octave.rb [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+              --with-hdf5-includedir=#{Formula["hdf5"].opt_include}
```
Comment (ilovezfs) [inline Formula/octave.rb]: doesn't find this on its own?

File: Formula/octave.rb [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+              --with-hdf5-libdir=#{Formula["hdf5"].opt_lib}
```
Comment (ilovezfs) [inline Formula/octave.rb]: doesn't find this on its own?

File: Formula/octave.rb [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+  depends_on "suite-sparse"
+  depends_on "transfig"
+  depends_on "veclibfort"
+
+  # If GraphicsMagick was built from source, it is possible that it was
+  # done to change quantum depth. If so, our Octave bottles are no good.
+  # https://github.com/Homebrew/homebrew-science/issues/2737
+  if build.with? "graphicsmagick"
+    def pour_bottle?
+      Tab.for_name("graphicsmagick").bottle?
```
Comment (ilovezfs) [inline Formula/octave.rb]: should be able to check the specific options.

**Comment by fxcoudert** [comment]
- bison (a recent one) is only needed for head build
- gawk is not necessary, awk works fine 4.2.1
- texinfo not needed as we don't build docs
- curl not needed
- metis is indirect dep, not needed
- system pcre and readline can be used
- flags manipulation is not necessary
- `--enable-link-all-dependencies` is crucial (otherwise fails to link)

**Comment by fxcoudert** [comment]
Removed X11 and fontconfig flags. Readded `pcre` because it needs the header.

**Comment by fxcoudert** [comment]
More fined-grained logic for pouring bottles based on `graphicsmagick` options.

**Comment by fxcoudert** [comment]
Build head with same options as release (thus defaulting to 64-bit indexes).

I believe this is ready to 🛳

**Review by ilovezfs — APPROVED** [review]

**Comment by schoeps** [comment]
Were experienced users of octave involved in these steps? If not, why didn't you contact previous contributors? I am a bit astonished that many (interesting) options have been removed, e.g. qt5 and more importantly openblas.  

Most scientists do not trust Apple's BLAS (and its results are not reproducibly on other platforms), so this octave should not be used for scientific research. In this regard was a reasonably choice to move it out of science :(

**Comment by ilovezfs** [comment]
@schoeps it does use veclibfort, but I don't mind putting back the openblas option. The qt5 option cannot return until it's able to use the main qt formula (currently at version 5.9.1). See discussion on https://github.com/Homebrew/homebrew-core/issues/15986.

**Comment by ilovezfs** [comment]
To put some numbers on this, in the last nine days since it was moved to core, there have been a total of 1,584 installs of the formula. We've had three people, including you, mention the qt GUI having been removed.

**Comment by MikeMcQuaid** [comment]
> I am a bit astonished that many (interesting) options have been removed, e.g. qt5 and more importantly openblas.

We are unable to support large numbers of options (and option combinations) to a reasonable degree as we are unable to test them in CI. We look at the options our users use using analytics and use that to decide both new defaults and what options to continue to support.

**Comment by schoeps** [comment]
@ilovezfs I think it's bit sad that you guys did this transition without involving former contributors. I do this in my free time, so I am not following the developments so closely. Some stuff could have been discussed beforehand.
* regarding qt: I personally do not use the GUI and many experienced octavers don't. However, it is a very frequently asked question on the octave mailing list and I was still aiming at fixing the qt problems. I assume that most of the 1,584 users are rather 'installers' than 'users'.
* regarding openblas: all dependencies must be adapted. I recall that you (?) thought about a better handling of blas libraries but this was cancelled, right?

**Comment by schoeps** [comment]
@MikeMcQuaid I understand that reducing options is easier for maintenance but then you might loose experienced users. If you are more interested in number than this is the obvious way to go. This is a bit sad as I will now have to set up my private science repository based on old formulas. At least I need some more control over my tool chain :(

**Comment by MikeMcQuaid** [comment]
> I do this in my free time, so I am not following the developments so closely.

We appreciate that and your work on Homebrew (genuinely) but it's worth remember: we're all doing this in our free time, too.

> I understand that reducing options is easier for maintenance but then you might loose experienced users.

The other side is every time a user tries to install an option that doesn't work we are also at risk of losing experienced users. I'd rather Homebrew tried to do less and was higher-quality overall.

> This is a bit sad as I will now have to set up my private science repository based on old formulas. At least I need some more control over my tool chain :(

I don't see this as sad, personally, because you still have the option to support formulae in this way for users that need to use it in this way. It's great that you're willing and able to support some options that are are not.

**Comment by ilovezfs** [comment]
@schoeps I apologize for not having pinged you during the transition.

>regarding qt: I personally do not use the GUI and many experienced octavers don't. However, it is a very frequently asked question on the octave mailing list and I was still aiming at fixing the qt problems.

The option is welcome to return as soon as it works with the current qt formula.

>I assume that most of the 1,584 users are rather 'installers' than 'users'.

Of the 1,584 installs, 1,486 were install-on-request events.

>regarding openblas: all dependencies must be adapted.

Due to https://github.com/Homebrew/brew/pull/2482/commits/3011cbac34334c37eba20cbf06fe7942522602f6 this won't be feasible unless we vendor all of the dependencies into the formula, have a separate openblas version for each, or make openblas the default. Which of the dependencies need to be built against openblas for it to work right?

>I recall that you (?) thought about a better handling of blas libraries but this was cancelled, right?

@dpo was the one thinking about this.

**Comment by dpo** [comment]
> We look at the options our users use using analytics and use that to decide both new defaults and what options to continue to support.

@MikeMcQuaid I was wondering about that. Would you mind giving me access to detailed the analytics results? I'm especially curious about popular options.

**Comment by dpo** [comment]
> I recall that you (?) thought about a better handling of blas libraries but this was cancelled, right?
>> @dpo was the one thinking about this.

Still am, very much, if only I could find the time.

**Comment by schoeps** [comment]
>> @dpo was the one thinking about this.
>Still am, very much, if only I could find the time.

@dpo I think this is now even more important than before. Sadly, I cannot help you (since I do the homebrew internals not well enough) but I will gladly help you with testing etc.

**Comment by ilovezfs** [comment]
@schoeps @dpo I think what's needed is an OpenBLAS tap that has copies of all formulae from homebrew/core and homebrew/science that can use OpenBLAS on macOS.

**Comment by dpo** [comment]
@ilovezfs They're just all the formulas that use `veclibfort` or somehow link with `Accelerate`. Do you mean that they would link with `openblas` in the tap so users can access an entire toolchain linked against the same BLAS/LAPACK?

**Comment by ilovezfs** [comment]
@dpo yep. Note that third party taps can add bottles if they want to, so it could potentially be a nicer experience too if the tap goes that direction.

**Comment by ilovezfs** [comment]
@dpo in `man brew` see the `brew tap-pin` command.

**Comment by dpo** [comment]
That's surely an interesting possibility, especially if we provide bottles. Some of the formulae concerned are in core while others are in science though. Does that mean we'd need two taps? Also, is there any way to automate the process of updating a formula when the corresponding formula in core or science is updated?

**Comment by dpo** [comment]
In core, the following formulae are affected:
- arpack.rb (straightforward)
- octave.rb (straightforward)
- r.rb (straightforward)
- scalapack.rb (straightforward)
- suite-sparse.rb (straightforward)
- qrupdate.rb (striaghtforward)
- rubberband.rb (need to look more closely)
- scipy.rb (need to look more closely)
- viennacl.rb (current install seems too simplistic at first sight)

By the way, `r.rb` still gives the choice of Accelerate vs. OpenBLAS.

`man brew` says
```
 tap-pin tap
        Pin tap, prioritizing its formulae over core when formula names are supplied by the user.
```
so apparently, `tap-pin` only affects formulae in core. How about homebrew/science?

**Comment by ilovezfs** [comment]
>By the way, r.rb still gives the choice of Accelerate vs. OpenBLAS.

We'll be removing that and adding an audit to prevent its use as dependency within the organization.

>so apparently, tap-pin only affects formulae in core. How about homebrew/science

You'd need to test it. I don't know off hand but it probably overshadows them too.

>Also,  is there any way to automate the process of updating a formula when the  corresponding formula in core or science is updated?

You could automate it with a cron job or similar, and bump-formula-pr.

**Comment by dpo** [comment]
> We'll be removing that and adding an audit to prevent its use as dependency within the organization.

So you mean this wouldn't be an official tap?

**Comment by ilovezfs** [comment]
@dpo right. The audit barring an OpenBLAS dependency wouldn't apply to an OpenBLAS tap.

**Comment by schoeps** [comment]
Is that openblas tap meant to cure the problem (i.e. the user sets the default blas to openblas and gets transparently the new tab) or is it meant to shadow existing formulas? How is the new tap kept uptodate? I thought that @dpo is aiming at somwthing like ":blas".

**Comment by ilovezfs** [comment]
>Is that openblas tap meant to cure the problem (i.e. the user sets the default blas to openblas and gets transparently the new tab) or is it meant to shadow existing formulas?

It would shadow existing formulas.

>How is the new tap kept uptodate?

https://github.com/Homebrew/brew/blob/master/docs/How-to-Create-and-Maintain-a-Tap.md#how-to-create-and-maintain-a-tap

>I thought that @dpo is aiming at somwthing like ":blas".

That would probably mean the user would have to build everything from source if they're using the non-default BLAS.

**Comment by MikeMcQuaid** [comment]
To provide a bit of context here: we're trying to eliminate more options and particularly options that aren't widely used and formulae that depend on other formulae with options. This is because these have to be built from source and aren't testing through CI and as a result they break often. When things break in official taps they reflect poorly on Homebrew as a whole and increase our support burden which means we can't spend as much time focused on the default, majority Homebrew experience. We'd rather support a smaller subset of options better and allow third-party individuals, groups and taps to maintain things that they use.

**Comment by dpo** [comment]
@MikeMcQuaid That makes sense. A `:blas` extension would be nice but users choosing a non-default BLAS would still need to build from source. Creating a tap that offers uses an `openblas` toolchain isn't difficult. I just noticed CircleCI offers a small-scale free OSX option for open-source projects that might make sense to ship bottles.

@schoeps Would you have a little time to help here and there with such a tap (especially in relation to `octave`)?

**Comment by schoeps** [comment]
Ideally, for a different BLAS library one just changes a some hardlinks or one could even use [FlexiBLAS](http://www.netlib.org/lapack/lawnspdf/lawn284.pdf). So `:blas` seems to be a more long-lasting solution. However, what about the usage of `:mpi`? I recall that this is often passed to dependencies in `science`. Another tap?

@MikeMcQuaid what do you think about `:blas`? Depending on the implementation, we could ensure that all formula are tested with all blas backends and we could even create binaries, couldn't we?

@dpo: yes. I am on holidays right now, but we could aim at a code sprint end of august. However, I am concerned that we end up replicating more and more formulas (which need to be kept in sync). What about other software from core or `science` that are using blas? Wouldn't they break?

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid what do you think about :blas? Depending on the implementation, we could ensure that all formula are tested with all BLAS backends and we could even create bottles, couldn't we?

@schoeps We just don't have the resources to build formulae more than once, unfortunately and we don't want to maintain multiple versions of the same formula. I realise this scope cutting is frustrating and I apologise for that; what we really want to do is figure out how to have good defaults for creating a tap so that e.g. people can create bottles with free resources.

**Comment by dpo** [comment]
@ilovezfs I'm getting audit errors saying that `openblas` is provided by macOS (it isn't): https://travis-ci.org/dpo/homebrew-openblas/builds/262455290#L3061. Will there be a way to deactivate `openblas`-related audits in third-party taps?

**Comment by ilovezfs** [comment]
@dpo that is a `--new-formula` audit so it just has to be ignored when the formula is initially accepted into the tap, and then you shouldn't hear from it again.

**Comment by dpo** [comment]
Ok, thanks. As a side note, I find it misleading to say that OpenBLAS is provided by macOS. macOS provides another BLAS implementation in Accelerate, not OpenBLAS. Perhaps
```
Dependency 'openblas' may be unnecessary as macOS already provides optimized blas and lapack; try to build this formula without it.
```

**Comment by ilovezfs** [comment]
@dpo it's an automatically generated message from
```

  keg_only :provided_by_osx,
           "macOS provides BLAS and LAPACK in the Accelerate framework"
```
so you can tweak that part of the wording if you want to open a PR.

**Comment by dpo** [comment]
Sorry, I was referring to the [audit message](https://github.com/Homebrew/brew/blob/1a30c165bfb82ae846e64c508f06455e0d3f285e/Library/Homebrew/dev-cmd/audit.rb#L460).

**Comment by ilovezfs** [comment]
@dpo Ah, yeah. I don't think there's anything we can do about that message without special casing it.

**Comment by MikeMcQuaid** [comment]
We could change the audit to use the reason string instead.

**Comment by dpo** [comment]
@ilovezfs @MikeMcQuaid https://github.com/dpo/homebrew-openblas

**Pull request closed without merging**